test/boost/mutation_reader_test.cc

```diff
@@ -3658,3 +3658,214 @@ SEASTAR_THREAD_TEST_CASE(test_clustering_order_merger_in_memory) {
+SEASTAR_THREAD_TEST_CASE(clustering_combined_reader_mutation_source_test) {
+    // The clustering combined reader (based on `clustering_order_reader_merger`) supports only
+    // single partition readers, but the mutation source test framework tests the provided source
+    // on multiple partitions. Furthermore, our reader doesn't support static rows or partition tombstones.
+    // We deal with this as follows:
+    // 1. we process all mutations provided by the framework, extracting unsupported fragments
+    //    (partition tombstones and static rows) into separate mutations, called ``bad mutations'' below.
+    //    Each bad mutation is used to create a separate reader.
+    // 2. The remaining mutations (``good mutations'') are sorted w.r.t their partition keys;
+    //    for each partition key, we use the set of mutations with this key to create our clustering combined reader.
+    //    The readers are then fed into a meta-reader `multi_partition_reader` which supports multi-partition queries,
+    //    calling the provided readers appropriately as the query goes through the partition range.
+    // 3. The ``bad mutation'' readers from step 1 and the reader from step 2 are combined together
+    //    using the generic combined reader.
+
+    // Multi partition reader from single partition readers.
+    // precondition: readers must be nonempty (they return a partition_start)
+    struct multi_partition_reader : public flat_mutation_reader::impl {
+        using container_t = std::map<dht::decorated_key, flat_mutation_reader, dht::decorated_key::less_comparator>;
+
+        container_t _readers;
+        container_t::iterator _it;
+
+        // invariants:
+        // 1. _it = end() or _it->first is not before the current partition range (_range)
+        // 2. _it->second did not return partition_end
+
+        // did we fetch a fragment from _it? (due to invariant 2, it wasn't partition end)
+        bool _inside_partition = false;
+
+        std::reference_wrapper<const dht::partition_range> _range;
+
+        multi_partition_reader(schema_ptr s, reader_permit permit,
+                container_t readers, const dht::partition_range& range)
+            : impl(std::move(s), std::move(permit))
+            , _readers(std::move(readers)), _range(range)
+        {
+            assert(!_readers.empty());
+
+            _it = _readers.begin();
+            // restore invariant 2
+            while (_it != _readers.end() && _range.get().before(_it->first, dht::ring_position_comparator(*_schema))) {
+                ++_it;
+            }
+        }
+
+        virtual future<> fill_buffer(db::timeout_clock::time_point timeout) override {
+            while (!is_buffer_full()) {
+                auto mfo = co_await next_fragment(timeout);
+                if (!mfo) {
+                    _end_of_stream = true;
+                    break;
+                }
+                push_mutation_fragment(std::move(*mfo));
+            }
+        }
```

Comment:
coroutines treat falling of the end as `co_return;`. `co_return x` translates to `return make_ready_future<T>(std::move(x))`, and `co_return;` translates to `return make_ready_future<>()`.

Note: this is for professionals only. Do not try falling off the end at home.

Reply:
It looks like for coroutines there's an implicit `co_return` inserted at the end, just like `return` for functions.
In case of the `future` coroutine type it looks like it returns `make_ready_future<>()` (not sure - but looks like it works)